libraries/AP_OSD/AP_OSD_ParamSetting.cpp

```diff
@@ -0,0 +1,418 @@
+    "", "", "ACRO", "", "", "GUIDE", "LOIT", "FOLLOW", "CLR_WP", "",
+    "ZZAG", "ZZ_SVWP", "COMP_LRN", "", "", "GPS_DIS", "RLY5", "RLY6", "STAB", "PHOLD",
+    "AHOLD", "FHOLD", "CIRCLE", "DRIFT", "", "", "STANDBY", "", "RCAM_CTL", "RCAM_OSD",
+    "VISO_CAL", "DISARM", "", "ZZ_Auto", "AIR", "", "", "", "", "",
+    "", "", "", "", "", "", "", "", "", "",
+    "KILLIMU1", "KILLIMU2", "CAM_MOD_TOG", "", "", "GPSYAW_DIS"
+};
+
+static const char* FLTMODES[] = {
+    "STAB", "ACRO", "ALTHOLD", "AUTO", "GUIDED", "LOIT", "RTL", "CIRC", "", "LAND",
+    "", "DRFT", "", "SPORT", "FLIP", "ATUN", "POSHLD", "BRAKE", "THROW", "AVD_ADSB",
+    "GUID_NOGPS", "SMRTRTL", "FLOHOLD", "FOLLOW", "ZIGZAG", "SYSID", "HELI_ARO"
+};
+
+static const char* FS_OPTIONS[] = {
+    "NONE", "CONT_RCFS", "CONT_GCSFS", "CONT_RC/GCSFS", "", "", "", "CONT_LAND", "",
+    "", "", "", "", "", "", "CONT_CTRL_GCS", "", "", "CONTNUE"
+};
+
+static const char* THR_FS_ACT[] = {
+    "NONE", "RTL", "CONT", "LAND", "SRTL_RTL", "SRTL_LAND"
+};
+
+static const char* FS_ACT[] = {
+    "NONE", "LAND", "RTL", "SRTL_RTL", "SRTL_LAND", "TERM"
+};
+
+// copter parameters
+const AP_OSD_ParamSetting::ParamMetadata AP_OSD_ParamSetting::_param_metadata[OSD_PARAM_NUM_TYPES] = {
+    { -1, 31, 1, ARRAY_SIZE(SERIAL_PROTOCOL_VALUES), SERIAL_PROTOCOL_VALUES },  // OSD_PARAM_SERIAL_PROTOCOL
+    { 0, 132, 1, ARRAY_SIZE(SERVO_FUNCTIONS), SERVO_FUNCTIONS },                // OSD_PARAM_SERVO_FUNCTION
+    { 0, 105, 1, ARRAY_SIZE(AUX_OPTIONS), AUX_OPTIONS },                        // OSD_PARAM_AUX_FUNCTION
+    { 0, 23, 1,  ARRAY_SIZE(FLTMODES), FLTMODES },                              // OSD_PARAM_FLIGHT_MODE
+    { 0, 3, 1,   ARRAY_SIZE(FS_OPTIONS), FS_OPTIONS },                          // OSD_PARAM_FAILSAFE_ACTION
+    { 0, 5, 1,   ARRAY_SIZE(FS_ACT), FS_ACT },                                  // OSD_PARAM_FAILSAFE_ACTION_1
+    { 0, 5, 1,   ARRAY_SIZE(THR_FS_ACT), THR_FS_ACT },                          // OSD_PARAM_FAILSAFE_ACTION_2
+};
+
+#else
+const AP_OSD_ParamSetting::ParamMetadata AP_OSD_ParamSetting::_param_metadata[OSD_PARAM_NUM_TYPES] = {};
+#endif
+
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_OSD_ParamSetting::AP_OSD_ParamSetting(uint8_t param_number, bool _enabled, uint8_t x, uint8_t y,  int16_t key, int8_t idx, int32_t group, float min, float max, float incr)
+    : AP_OSD_Setting(_enabled, x, y), _param_number(param_number)
+{
+    _param_group = group;
+    _param_idx = idx;
+    _param_key = key;
+}
+
+// update the contained parameter
+void AP_OSD_ParamSetting::update()
+{
+    if (PARAM_TOKEN_INDEX(_current_token) == PARAM_COMPOSITE_INDEX(_param_key, _param_idx, _param_group) && _param_key >= 0) {
+        return;
+    }
+    // if a parameter was configured then use that
+    _current_token = AP_Param::ParamToken {};
+    // surely there is a more efficient way than brute-force search
+    for (_param = AP_Param::first(&_current_token, &_param_type);
```

Comment:
ok with cpu cost? only called when user makes a change?

Reply:
Yes cached and only changed when the user changes it